test/functional/pdo_sqlsrv/MsCommon_mid-refactor.inc

```diff
@@ -0,0 +1,1833 @@
+        $createSql = "CREATE TABLE $tbname ($colDef)";
+        $numRows = $conn->exec($createSql);
+        return $numRows;
+    } catch (PDOException $e) {
+        var_dump($e->errorInfo);
+    } catch (Exception $e) {
+        var_dump($e->errorInfo);
+        exit;
+    }
+}
+
+
+/**
+ * class for encapsulating optional parameters for PDOStatement::bindParam
+ */
+class BindParamOp
+{
+    public $parameter;  //parameter identifier. Parameter name or 1-based index of the parameter
+    public $variable;   //name of the PHP variable to bind
+    public $pdoType;    //PDO_PARAM_ type specified for PDOStatement::bindParam
+    public $length;     //length specified for PDOStatement::bindParam
+    public $options;    //options specified for PDOStatement::bindParam
+
+    public function __construct($parameter, $variable, $pdoType = null, $length = null, $options = null)
+    {
+        $this->parameter = $parameter;
+        $this->variable = $variable;
+        $this->pdoType = $pdoType;
+        $this->length = $length;
+        $this->options = $options;
+    }
+    /**
+     * @param  object  $stmt : PDO Statement object
+     */
+    public function bindWithOp($stmt)
+    {
+        // get the constant values of pdoType, and option
+        $pdoType = null;
+        $options = null;
+        if (!is_null($this->pdoType)) {
+            $pdoType = constant($this->pdoType);
+        }
+        if (!is_null($this->options)) {
+            $options = constant($this->options);
+        }
+        $stmt->bindParam($this->parameter, $this->variable, $pdoType, $this->length, $options);
+    }
+}
+
+
+/**
+ * Insert a row into a table
+ * @param  object  $conn : PDO connection object
+ * @param  string  $tbname : name of the table for the row to be inserted
+ * @param  array  $inputs : an associative array column name and its value; value may be a literal value or a BindParamOp object
+ * @param  string  $api : PDO_SQLSRV API used for executing the insert query
+ *                      accepted values: "exec", "query", "prepareDirect", "prepareExecuteBind", "prepareBindParam"
+ * @param  bool  $r : true if the row was successfully inserted, otherwise false. Default value is null to make this parameter optional.
+ * @return object PDOStatement object of the insert statement
+ */
+function insertRow($conn, $tbname, $inputs, $api = null, &$r = null)
+{
+    try {
+        // set the default API being called depending on if column is encrypted
+        if (is_null($api)) {
+            if (!isColEncrypted()) {
+                $api = "query";
+            } else {
+                $api = "prepareBindParam";
+            }
+        }
+        $stmt = null;
+        if (!isColEncrypted() && $api != "prepareExecuteBind" && $api != "prepareBindParam") {
+            $insertSql = getInsertSqlComplete($tbname, $inputs);
+            switch ($api) {
+                case "exec":
+                    $conn->exec($insertSql);
```

Comment:
the return value for exec() is the rowcount. I don't see how I can return that (otherwise I can have a mix type of return values). $r is needed for tests that want to handle the error separately. If the errmode is exception, an exception will be thrown and it'll be caught in the function. $stmt is initialized as null, so if it's not assigned, null will be returned.

Reply:
Yes I understand, but what if something goes wrong in exec()? Like a syntax error of the t-sql statement

Reply:
The return value for exec doesn't tell you if anything went wrong though. It'll be up to the try catch block if anything happens.